Five a day fix for users with less than five suggestions available or…

mixolidia opened the pull request:
… not onboarded.

Add show_five? method to to lib/exercism/user.rb to show five a day on navbar only when user is not a guest, is onboarded and has five suggestions available.
Add show_five? method to lib/exercism/guest.rb as false for guests.
Add show_five? method to lib/app/views/application/_nav.erb.

kytrinyx commented on lib/app/views/application/_nav.erb:
```diff
@@ -26,11 +26,13 @@
-      <% if current_user.sees_exercises? %>
-        <li><a href="/inbox">Exercises</a></li>
-      <% end %>
+    <% if current_user.show_five? == true %>
```
This comparison is the equivalent of `true == true` or `false == true`. Do you see how it can be simplified?

kytrinyx commented on lib/exercism/user.rb:
```diff
@@ -124,11 +124,20 @@ def count_existing_five_a_day
+    else
+      true
+    end
+  end
```
Here, too, you have a conditional that compares booleans to return booleans. These can usually be simplified so that you return the booleans directly.

kytrinyx commented on lib/exercism/user.rb:
```diff
@@ -124,11 +124,20 @@ def count_existing_five_a_day
   end
 
+  def show_five?
+    if guest? && !onboarded?
```
A logged in user is never a guest.

kytrinyx commented on lib/exercism/user.rb:
```diff
@@ -124,11 +124,20 @@ def count_existing_five_a_day
+  def show_five?
+    if guest? && !onboarded?
+      false
+    elsif five_a_day_exercises.count + count_existing_five_a_day != 5
```
How would you describe `five_a_day_exercises.count + count_existing_five_a_day` in English? Should we extract it into a method for the sake of readability?

mixolidia commented:
Hi @kytrinyx 
I created a new method called `five_available?` to keep `five_a_day_exercises.count + count_existing_five_a_day != 5` in and returns true or false depending.

I changed the name of show_five? to show_five_suggestions? and updated it where it was needed. I also replaced `five_a_day_exercises.count + count_existing_five_a_day != 5` on the last else. I think this makes it simpler between the names of the methods and breaking everything out. What do you think?

I was just double checking truths everywhere just to make sure. I know it was silly.

cc: @sguermond

kytrinyx commented:
Yeah, this looks much better. I know the feeling about truths, but usually better names is the way to go if you're feeling worried or uncertain.

kytrinyx commented on lib/exercism/user.rb:
```diff
@@ -124,11 +124,28 @@ def count_existing_five_a_day
 
+  def show_five_suggestions?
+    if guest?
+      false
```
This clause can be deleted. guest? is always false inside the user class.

kytrinyx commented on lib/exercism/user.rb:
```diff
@@ -124,11 +124,28 @@ def count_existing_five_a_day
+      false
+    else
+      five_available?
+    end
```
A simpler version of this whole method would be:

```
onboarded? && five_available?
```

kytrinyx commented on lib/exercism/user.rb:
```diff
@@ -124,11 +124,28 @@ def count_existing_five_a_day
+    false
+  else
+    true
+  end
```
A simpler version of this method would be:

```
five_a_day_exercises.count + count_existing_five_a_day == 5
```

mixolidia commented:
Hi @kytrinyx,
Thanks for showing me. I made the updates to show_five_suggestions? and five_available? & pushed.

cc: @sguermond

kytrinyx commented on lib/exercism/user.rb:
```diff
@@ -124,11 +124,22 @@ def count_existing_five_a_day
+  def show_five_suggestions?
+    if onboarded? && five_available?
+      true
+    end
```
This is the equivalent of:

```
if true
  true
else
  nil
end
```

This method implementation should be a one-liner:

```
def show_five_suggestions?
  onboarded? && five_available?
end
```

That way the return value will be either true or false, not true or nil.

kytrinyx commented on lib/exercism/user.rb:
```diff
@@ -124,11 +124,22 @@ def count_existing_five_a_day
+  if five_a_day_exercises.count + count_existing_five_a_day == 5
+    true
+  end
+end
```
Same deal here: don't check a boolean to return a boolean.

kytrinyx commented:
I amended the commit locally and merged it in 7efcc2bc6d48b72caaf02748adb8c8a7c1e06dae

mixolidia commented:
Hi @kytrinyx,

I was at Open Source & Feelings Friday & Saturday. I didn't get to my computer until now.

kytrinyx commented:
Nice! I heard good things about the conference.

mixolidia commented:
Yep! It was really good. They will have videos of the talks up soon. Might want to keep an eye on [OSFeels Twitter feed](https://twitter.com/OSFeels).
